### PR TITLE
Refine TypedMutableSequence.__getitem__ error type, add magicgui tests

### PR DESCRIPTION
**Pull request by tlambert03**
closes #1956  (supersedes it ... but thank you @brisvag for catching that very important error!)
This makes the error type of `TypedMutableSequence.__getitem__(key)` a little more refined:
1. If `type(key)` is in the `TypedMutableSequence._lookup` dict provided by the user (such as `str` in our `LayerList`), then if the item corresponding to `key` is not in the list, it raises `KeyError`.
2. If `key` is a regular integer (and `int` is not in `self._lookup` ... which would be strange anyway), it raises `IndexError`
3.  `TypedMutableSequence.index(key)` still raises a `ValueError` if `key` is not in the list.

also added a few basic magicgui tests to catch something like this in the future, and added magicgui to test requirements.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by tlambert03** [comment]
not sure why, but was getting segfault on one specific test on mac CI (but not locally).  just gonna skip it for now.

File: napari/utils/_tests/test_magicgui.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    viewer = make_test_viewer()
+    viewer.window.add_dock_widget(add_labels.Gui())
+    assert len(viewer.layers) == 0
+    add_labels()  # should add a new layer to the list
+    assert len(viewer.layers) == 1
+    assert viewer.layers[0].name == 'add_labels result'
+    assert isinstance(viewer.layers[0], Image)
+
+
+@pytest.mark.skipif(sys.platform == "darwin", reason="segfault on mac CI")
```
Comment (sofroniewn) [inline napari/utils/_tests/test_magicgui.py:47]: Was this a known issue, or unexpected?

**Review by sofroniewn — APPROVED** [review]
Ok this works great for me, have given it a go and it fixes the issue. I asked a question about the mac CI seg fault, but no need for it to hold up merge I will merge now and try and add @brisvag as a coauthor. Thanks for getting this started with #1956 @brisvag and thanks for finishing it up @tlambert03!!

File: napari/utils/_tests/test_magicgui.py (hunk 1 above)
Comment (sofroniewn) [reply napari/utils/_tests/test_magicgui.py:47]: > not sure why, but was getting segfault on one specific test on mac CI (but not locally). just gonna skip it for now.

Ah just saw this, sorry missed it during my review. That's fair enough, we can revisit later

Comment (sofroniewn) [reply napari/utils/_tests/test_magicgui.py:47]: We're actually getting some fails on window here too, see https://cirrus-ci.com/task/5190124025675776 on master right now

```
napari/utils/_tests/test_magicgui.py::test_magicgui_returns_layer_tuple Windows fatal exception: access violation
```

I think it's fine to wait till after `0.4.2` release to fix, but wanted to flag so we don't forget. It appears to happen non-deterministically

Comment (tlambert03) [reply napari/utils/_tests/test_magicgui.py:47]: ok, should probably just remove that test for now

**Pull request merged**